compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs

```diff
@@ -921,22 +921,19 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             let ty = <dyn AstConv<'_>>::ast_ty_to_ty(self, ty);
             let bound_vars = self.tcx.late_bound_vars(fn_id);
             let ty = self.tcx.erase_late_bound_regions(Binder::bind_with_vars(ty, bound_vars));
-            let ty = self.normalize_associated_types_in(expr.span, ty);
             let ty = match self.tcx.asyncness(fn_id.owner) {
-                hir::IsAsync::Async => {
-                    let infcx = self.tcx.infer_ctxt().build();
-                    infcx
-                        .get_impl_future_output_ty(ty)
-                        .unwrap_or_else(|| {
-                            span_bug!(
-                                fn_decl.output.span(),
-                                "failed to get output type of async function"
-                            )
-                        })
-                        .skip_binder()
-                }
+                hir::IsAsync::Async => self
+                    .get_impl_future_output_ty(ty)
+                    .unwrap_or_else(|| {
+                        span_bug!(
+                            fn_decl.output.span(),
+                            "failed to get output type of async function"
+                        )
+                    })
+                    .skip_binder(),
                 hir::IsAsync::NotAsync => ty,
             };
+            let ty = self.normalize_associated_types_in(expr.span, ty);
```

Comment:
For the record, moving this normalize call is what allows the UI test I committed to properly emit:

help: you might have meant to return this value
   |
LL |         return 1i32;
   |         ++++++     +